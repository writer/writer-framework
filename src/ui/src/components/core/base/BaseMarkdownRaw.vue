<template>
	<div v-dompurify-html="rawMarkdown" class="markdown"></div>
</template>

<script setup lang="ts">
defineProps({
	rawMarkdown: { type: String, required: true },
});
</script>

<style scoped>
.markdown {
	font-size: 0.875rem;
}

.markdown:deep() h1,
.markdown:deep() h2,
.markdown:deep() h3,
.markdown:deep() h4 {
	margin: 0;
	color: var(--primaryTextColor);
}

.markdown:deep() h1 {
	font-size: 1.5rem;
	letter-spacing: 0.1875rem;
	font-weight: 600;
	line-height: 162.5%;
}

.markdown:deep() h2 {
	font-size: 1.5rem;
	font-weight: 500;
	line-height: 120%;
}

.markdown:deep() h3 {
	font-size: 1.25rem;
	font-weight: 500;
	line-height: 120%;
}

.markdown:deep() h4 {
	font-weight: 700;
	font-size: 1rem;
	line-height: 247.5%;
}

.markdown:deep() ul,
.markdown:deep() ol {
	padding: 0;
	padding-inline-start: 0;
	margin-block-start: 0;
}

.markdown:deep() li {
	margin: 0 0 0 32px;
}

.markdown:deep() hr {
	border: none;
	border-top: 1px solid var(--separatorColor);
}

.markdown:deep() pre {
	background-color: var(--separatorColor);
	font-family: monospace;
	padding: 8px;
}

.markdown:deep() code {
	background-color: var(--separatorColor);
	font-family: monospace;
	padding: 2px;
}

.markdown:deep() pre > code {
	background-color: unset;
}

.markdown:deep() table {
	border-collapse: collapse;
}

.markdown:deep() th {
	padding: 8px;
	background: var(--separatorColor);
	border: 1px solid var(--separatorColor);
}

.markdown:deep() td {
	padding: 8px;
	border: 1px solid var(--separatorColor);
}

.markdown:deep() blockquote {
	padding: 8px;
	border-left: 5px solid var(--separatorColor);
}
</style>
