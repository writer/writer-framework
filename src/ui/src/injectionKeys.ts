import { ComputedRef, InjectionKey, Ref, VNode } from "vue";
import {
	BuilderManager,
	CollaborationManager,
	Component,
	Core,
	InstancePath,
	InstancePathItem,
	NotesManager,
	SecretsManager,
} from "./writerTypes";

export default {
	core: Symbol() as InjectionKey<Core>,
	builderManager: Symbol() as InjectionKey<BuilderManager>,
	notesManager: Symbol() as InjectionKey<NotesManager>,
	collaborationManager: Symbol() as InjectionKey<CollaborationManager>,
	secretsManager: Symbol() as InjectionKey<SecretsManager>,
	evaluatedFields: Symbol() as InjectionKey<Record<string, ComputedRef<any>>>,
	componentId: Symbol() as InjectionKey<Component["id"]>,
	isBeingEdited: Symbol() as InjectionKey<Ref<boolean>>,
	isDisabled: Symbol() as InjectionKey<Ref<boolean>>,
	getChildrenVNodes: Symbol() as InjectionKey<
		(instanceNumber?: InstancePathItem["instanceNumber"]) => VNode[]
	>,
	renderProxiedComponent: Symbol() as InjectionKey<
		(
			componentId: Component["id"],
			instanceNumber: InstancePathItem["instanceNumber"],
			ext?: { class?: string[]; contextSlot?: string },
		) => VNode
	>,
	instancePath: Symbol() as InjectionKey<InstancePath>,
	flattenedInstancePath: Symbol() as InjectionKey<string>,
	instanceData: Symbol() as InjectionKey<Ref[]>,
};
