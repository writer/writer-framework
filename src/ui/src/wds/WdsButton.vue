<template>
	<button class="WdsButton colorTransformer" :class="className" role="button">
		<slot></slot>
	</button>
</template>

<script setup lang="ts">
import { computed, PropType } from "vue";

/** See Variants on [Figma](https://www.figma.com/design/jgLDtwVwg3hReC1t4Vw20D/WDS-Writer-Design-System?node-id=67-701) */
type WdsButtonVariant =
	| "primary"
	| "secondary"
	| "tertiary"
	| "special"
	| "neutral";

/** See Sizes on [Figma](https://www.figma.com/design/jgLDtwVwg3hReC1t4Vw20D/WDS-Writer-Design-System?node-id=67-701) */
type WdsButtonSize = "big" | "small" | "icon" | "smallIcon";

const props = defineProps({
	variant: { type: String as PropType<WdsButtonVariant>, default: "primary" },
	size: { type: String as PropType<WdsButtonSize>, default: "big" },
});

const className = computed(() => [
	"WdsButton",
	`WdsButton--${props.variant}`,
	`WdsButton--${props.size}`,
]);
</script>

<style scoped>
@import "@/renderer/colorTransformations.css";

.WdsButton {
	width: fit-content;
	max-width: 100%;
	display: flex;
	align-items: center;
	justify-content: center;
	gap: 8px;
	border-radius: 300px;
	font-weight: 600;
	font-size: 0.875rem;
	cursor: pointer;
	box-shadow: var(--buttonShadow);
	outline: none;
	border-style: solid;
	border-width: 1px;
}

/* VARIANTS */

/* VARIANTS -- primary */

.WdsButton--primary {
	color: var(--buttonTextColor);
	background: var(--buttonColor);
	border-color: var(--buttonColor);
}

.WdsButton--primary:hover,
.WdsButton--primary:focus {
	border-color: var(--intensifiedButtonColor);
	background: var(--intensifiedButtonColor);
}
.WdsButton--primary:disabled {
	border-color: var(--wdsColorBlue6);
	background-color: var(--wdsColorBlue6);
	opacity: 40%;
}

/* VARIANTS -- secondary */

.WdsButton--secondary {
	color: var(--buttonTextColor);
	background: var(--wdsColorBlack);
	border-color: var(--wdsColorBlack);
}

.WdsButton--secondary:hover,
.WdsButton--secondary:focus {
	border-color: var(--wdsColorGray6);
	background: var(--wdsColorGray6);
}
.WdsButton--secondary:disabled {
	border-color: var(--wdsColorGray6);
	background: var(--wdsColorGray6);
	opacity: 40%;
}

/* VARIANTS -- tertiary */

.WdsButton--tertiary {
	color: var(--wdsColorBlack);
	background: var(--wdsColorWhite);
	border-color: var(--wdsColorGray2);
}

.WdsButton--tertiary:hover,
.WdsButton--tertiary:focus {
	color: var(--wdsColorGray4);
}
.WdsButton--tertiary:disabled {
	color: var(--wdsColorGray4);
	opacity: 50%;
}

/* VARIANTS -- special */

.WdsButton--special {
	color: var(--wdsColorBlue5);
	background: var(--wdsColorBlue2);
	border-color: var(--wdsColorBlue2);
}

.WdsButton--special:hover,
.WdsButton--special:focus {
	border-color: var(--wdsColorBlue3);
	background: var(--wdsColorBlue3);
}
.WdsButton--special:disabled {
	border-color: var(--wdsColorBlue2);
	background-color: var(--wdsColorBlue2);
	opacity: 0.4;
}

/* VARIANTS -- neutral (WDS "White" equivalent) */

.WdsButton--neutral {
	border: none;
	box-shadow: none;
	height: fit-content;
	background: unset;
	margin: 0;
	padding: 0;
}

.WdsButton--neutral:focus,
.WdsButton--neutral:hover {
	color: unset;
	border: none;
	box-shadow: none;
	background: var(--builderSubtleSeparatorColor);
}

.WdsButton--neutral:disabled {
	opacity: 0.4;
}

/* SIZES */

.WdsButton--big {
	padding: 10px 20px 10px 20px;
}

.WdsButton--small {
	height: 32px;
	padding: 4px 16px 4px 16px;
}

.WdsButton--icon {
	height: 40px;
	width: 40px;
}

.WdsButton--smallIcon {
	height: 32px;
	width: 32px;
}
</style>
