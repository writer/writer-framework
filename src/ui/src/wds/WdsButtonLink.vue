<template>
	<button
		class="WdsButtonLink"
		:disabled="disbaled"
		:class="className"
		role="button"
	>
		<i v-if="leftIcon" class="material-symbols-outlined">{{ leftIcon }}</i>
		<span class="WdsButtonLink__text">{{ text }}</span>
		<i v-if="rightIcon" class="material-symbols-outlined">{{
			rightIcon
		}}</i>
	</button>
</template>

<script lang="ts">
export type WdsButtonLinkWeight = "default" | "semibold";
export type WdsButtonLinkVariant = "primary" | "secondary";
</script>

<script setup lang="ts">
import { computed, PropType } from "vue";

const props = defineProps({
	variant: {
		type: String as PropType<WdsButtonLinkVariant>,
		default: "primary",
	},
	weight: {
		type: String as PropType<WdsButtonLinkWeight>,
		default: "default",
	},
	leftIcon: { type: String, required: false, default: undefined },
	rightIcon: { type: String, required: false, default: undefined },
	text: { type: String, required: true },
	disbaled: { type: Boolean },
});
const className = computed(() => [
	`WdsButtonLink--variant-${props.variant}`,
	`WdsButtonLink--weight-${props.weight}`,
]);
</script>

<style scoped>
.WdsButtonLink {
	background-color: transparent;
	border: none;
	text-align: left;
	display: flex;
	align-items: center;
	gap: 4px;

	cursor: pointer;
}
.WdsButtonLink:disabled {
	cursor: not-allowed;
	opacity: 40%;
}

.WdsButtonLink__text {
	text-overflow: ellipsis;
	white-space: nowrap;
	overflow: hidden;
	display: block;
}

.WdsButtonLink:hover .WdsButtonLink__text {
	text-decoration: underline;
}

.WdsButtonLink--variant-primary {
	color: var(--wdsColorBlue5);
}
.WdsButtonLink--variant-secondary {
	color: var(--wdsColorBlack);
}
</style>
