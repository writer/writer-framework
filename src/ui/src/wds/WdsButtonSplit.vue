<script lang="ts">
/** See Variants on [Figma](https://www.figma.com/design/jgLDtwVwg3hReC1t4Vw20D/.WDS-Writer-Design-System?node-id=8523-6185&t=wftDc1qTFGQKX8rM-4) */
export type WdsButtonSplitVariant =
	| "primary"
	| "secondary"
	| "tertiary"
	| "special";

export type WdsButtonSplitSize = "default" | "small";
</script>

<script setup lang="ts">
import BaseTransitionSlideFade from "@/components/core/base/BaseTransitionSlideFade.vue";
import { useFloating, offset } from "@floating-ui/vue";
import { computed, PropType, ref, useTemplateRef } from "vue";

const props = defineProps({
	variant: {
		type: String as PropType<WdsButtonSplitVariant>,
		default: "primary",
	},
	size: { type: String as PropType<WdsButtonSplitSize>, default: "default" },
	disabled: { type: Boolean, required: false },
});

const classNames = computed(() => [
	`WdsButtonSplit--size-${props.size}`,
	`WdsButtonSplit--variant-${props.variant}`,
	props.disabled ? `WdsButtonSplit--disabled` : undefined,
]);

const emits = defineEmits({
	mainClick: () => true,
	dropdownOpen: () => true,
	dropdownClose: () => true,
});

const btn = useTemplateRef("btn");
const dropdown = useTemplateRef("dropdown");

const isDropdownOpen = ref(false);

const { floatingStyles } = useFloating(btn, dropdown, {
	placement: "bottom-end",
	middleware: [offset(12)],
});

function toggleDropdown(value = !isDropdownOpen.value) {
	isDropdownOpen.value = value;
	isDropdownOpen.value ? emits("dropdownOpen") : emits("dropdownClose");
}

defineExpose({ toggleDropdown });
</script>

<template>
	<div ref="btn" class="WdsButtonSplit" :class="classNames">
		<button
			type="button"
			class="WdsButtonSplit__main"
			:disabled="disabled"
			@click="$emit('mainClick')"
		>
			<slot name="button" />
		</button>
		<hr class="WdsButtonSplit__divider" />
		<button
			class="WdsButtonSplit__dropdownTrigger"
			type="button"
			:disabled="disabled"
			@click.capture="toggleDropdown()"
		>
			<i class="material-symbols-outlined">{{
				isDropdownOpen ? "keyboard_arrow_up" : "keyboard_arrow_down"
			}}</i>
		</button>
		<BaseTransitionSlideFade>
			<div v-if="isDropdownOpen" ref="dropdown" :style="floatingStyles">
				<slot name="dropdown" />
			</div>
		</BaseTransitionSlideFade>
	</div>
</template>

<style lang="css" scoped>
.WdsButtonSplit {
	position: relative;
	display: flex;
	border-radius: 300px;
	box-shadow: var(--buttonShadow);
	border: 1px solid transparent;
	align-items: center;
	height: 40px;
	gap: 8px;
	padding-left: 12px;
	padding-right: 12px;
}

.WdsButtonSplit--disabled {
	opacity: 40%;
}
.WdsButtonSplit--disabled .WdsButtonSplit__dropdownTrigger,
.WdsButtonSplit--disabled .WdsButtonSplit__main {
	cursor: not-allowed;
}

.WdsButtonSplit__dropdownTrigger,
.WdsButtonSplit__main {
	background-color: transparent;
	border: none;
	cursor: pointer;
}

.WdsButtonSplit__main {
	display: flex;
	align-items: center;
	justify-content: flex-start;
	gap: 8px;
	font-weight: 600;
}
.WdsButtonSplit__dropdownTrigger {
	display: flex;
	align-items: center;
	justify-content: center;
	font-size: 20px;
	border-radius: 50%;
	height: 20px;
	width: 20px;
}
.WdsButtonSplit__divider {
	display: block;
	width: 1px;
	height: 100%;
	border: none;
	height: 24px;
}

/* variant - primary */

.WdsButtonSplit--variant-primary .WdsButtonSplit__divider {
	background: var(--wdsColorBlue4);
}
.WdsButtonSplit--variant-primary,
.WdsButtonSplit--variant-primary .WdsButtonSplit__dropdownTrigger {
	color: var(--wdsColorWhite);
	background: var(--wdsColorBlue5);
}
.WdsButtonSplit--variant-primary:not(.WdsButtonSplit--disabled):hover,
.WdsButtonSplit--variant-primary:not(.WdsButtonSplit--disabled)
	.WdsButtonSplit__dropdownTrigger:hover {
	background: var(--wdsColorBlue6);
}

/* variant - secondary */

.WdsButtonSplit--variant-secondary .WdsButtonSplit__divider {
	background: var(--wdsColorGray5);
}
.WdsButtonSplit--variant-secondary,
.WdsButtonSplit--variant-secondary .WdsButtonSplit__dropdownTrigger {
	color: var(--wdsColorWhite);
	background: var(--wdsColorBlack);
}
.WdsButtonSplit--variant-secondary:not(.WdsButtonSplit--disabled):hover,
.WdsButtonSplit--variant-secondary:not(.WdsButtonSplit--disabled)
	.WdsButtonSplit__dropdownTrigger:hover {
	background: var(--wdsColorGray6);
}

/* variant - tertiary */

.WdsButtonSplit--variant-tertiary .WdsButtonSplit__divider {
	background: var(--wdsColorGray2);
}
.WdsButtonSplit--variant-tertiary,
.WdsButtonSplit--variant-tertiary .WdsButtonSplit__dropdownTrigger {
	color: var(--wdsColorBlack);
	background: var(--wdsColorWhite);
}
.WdsButtonSplit--variant-tertiary:not(.WdsButtonSplit--disabled):hover,
.WdsButtonSplit--variant-tertiary:not(.WdsButtonSplit--disabled)
	.WdsButtonSplit__dropdownTrigger:hover {
	color: var(--wdsColorGray4);
}

/* variant - special */

.WdsButtonSplit--variant-special .WdsButtonSplit__divider {
	background: var(--wdsColorBlue5);
}
.WdsButtonSplit--variant-special,
.WdsButtonSplit--variant-special .WdsButtonSplit__dropdownTrigger {
	color: var(--wdsColorBlue5);
	background: var(--wdsColorBlue2);
}
.WdsButtonSplit--variant-special:not(.WdsButtonSplit--disabled):hover,
.WdsButtonSplit--variant-special:not(.WdsButtonSplit--disabled)
	.WdsButtonSplit__dropdownTrigger:hover {
	background: var(--wdsColorBlue3);
}

/* size */

.WdsButtonSplit--size-default .WdsButtonSplit__main {
	font-size: 14px;
}

.WdsButtonSplit--size-small .WdsButtonSplit__main {
	font-size: 12px;
}
</style>
