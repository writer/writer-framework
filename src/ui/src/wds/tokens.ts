import type { CSSProperties } from "vue";

/**
 * Colors from [Writer's design system](https://www.figma.com/design/jgLDtwVwg3hReC1t4Vw20D/.WDS-Writer-Design-System?node-id=1-2)
 */
export enum WdsColor {
	White = "#ffffff",
	Black = "#000000",

	Blue1 = "#f3f5ff",
	Blue2 = "#e4e9ff",
	Blue3 = "#bfcbff",
	Blue4 = "#6985ff",
	Blue5 = "#5551ff",
	Blue6 = "#4a46da",

	Gray0 = "#fbfbfd",
	Gray1 = "#f5f5f9",
	Gray2 = "#e4e7ed",
	Gray3 = "#d2d4d7",
	Gray4 = "#828282",
	Gray5 = "#4f4f4f",
	Gray6 = "#333333",

	Purple1 = "#f6effd",
	Purple2 = "#ede2ff",
	Purple3 = "#d4b2f7",
	Purple4 = "#a95ef8",
	Purple5 = "#721cc9",

	Orange1 = "#fff4f1",
	Orange2 = "#ffcfc2",
	Orange4 = "#ff8866",
	Orange5 = "#ff643c",

	Green1 = "#f2fffb",
	Green2 = "#d4fff2",
	Green3 = "#a9f9e1",
	Green5 = "#3bdcab",
	Green6 = "#078660",
}

/**
 * Shadows from [Writer's design system](https://www.figma.com/design/jgLDtwVwg3hReC1t4Vw20D/.WDS-Writer-Design-System?node-id=1-14)
 */
export enum WdsShadow {
	Box = "0px 2px 0px 0px #f3f3f3",
	Menu = "0px 1px 8px 0px #bfcbff40",
	Large = "0px 3px 40px 0px #acb9dc66",
}

export const WDS_CSS_PROPERTIES = Object.freeze<CSSProperties>(
	Object.fromEntries([
		...Object.entries(WdsColor).map(([key, value]) => [
			`--wdsColor${key}`,
			value,
		]),
		...Object.entries(WdsShadow).map(([key, value]) => [
			`--wdsShadow${key}`,
			value,
		]),
	]),
);
