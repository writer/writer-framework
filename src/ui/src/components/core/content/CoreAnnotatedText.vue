<template>
	<div v-if="shouldDisplay" class="CoreAnnotatedText">
		<BaseMarkdownRaw
			v-if="useMarkdown"
			:raw-markdown="markdown"
			class="CoreAnnotatedText__markdown"
		/>
		<template v-else>
			<BaseEmptiness v-if="isEmpty" :component-id="componentId" />
			<span v-for="(content, i) in text" :key="String(content) + i">
				<template v-if="typeof content === 'string'">{{
					content
				}}</template>
				<span
					v-if="Array.isArray(content)"
					class="CoreAnnotatedText__annotation"
					:style="{ backgroundColor: getAnnotationBgColor(content) }"
				>
					{{ content[0] }}
					<span
						v-if="content[1]"
						class="CoreAnnotatedText__annotation__subject"
					>
						{{ content[1] }}
					</span>
				</span>
			</span>
		</template>
		<SharedControlBar
			v-if="fields.copyButtons.value === 'yes'"
			:copy-raw-content="copyRawContent"
			:copy-structured-content="copyStructuredContent"
		/>
	</div>
</template>

<script lang="ts">
import {
	buttonColor,
	buttonTextColor,
	cssClasses,
	primaryTextColor,
} from "@/renderer/sharedStyleFields";
import SharedControlBar from "@/components/shared/SharedControlBar.vue";
import { WdsColor } from "@/wds/tokens";
import { validatorAnotatedText } from "@/constants/validators";
export default {
	writer: {
		name: "Annotated text",
		description: "Shows text with annotations",
		category: "Content",
		fields: {
			text: {
				name: "Annotated text",
				type: FieldType.Object,
				desc: "Value array with text/annotations. Must be a JSON string or a state reference to an array.",
				init: `["This ",["is", "Verb"]," some ",["annotated", "Adjective"], ["text", "Noun"]," for those of ",["you", "Pronoun"]," who ",["like", "Verb"]," this sort of ",["thing", "Noun"],". ","And here's a ",["word", "", "#faf"]," with a fancy background but no label."]`,
				validator: validatorAnotatedText,
			},
			referenceColor: {
				name: "Reference",
				desc: "The colour to be used as reference for chroma and luminance, and as the starting point for hue rotation.",
				type: FieldType.Color,
				default: WdsColor.Blue5,
				category: FieldCategory.Style,
			},
			seed: {
				name: "Seed value",
				desc: "Choose a different value to reshuffle colours.",
				type: FieldType.Number,
				default: "1",
				category: FieldCategory.Style,
			},
			rotateHue: {
				name: "Rotate hue",
				desc: "If active, rotates the hue depending on the content of the string. If turned off, the reference colour is always used.",
				type: FieldType.Text,
				options: {
					yes: "yes",
					no: "no",
				},
				default: "yes",
				category: FieldCategory.Style,
			},
			useMarkdown: {
				name: "Use Markdown",
				desc: "If active, the output will be sanitized; unsafe elements will be removed.",
				default: "no",
				type: FieldType.Text,
				options: {
					yes: "Yes",
					no: "No",
				},
			},
			copyButtons: {
				name: "Copy buttons",
				desc: "If active, adds a control bar with both copy text and JSON buttons.",
				type: FieldType.Text,
				options: {
					yes: "yes",
					no: "no",
				},
				default: "no",
				category: FieldCategory.Style,
			},
			buttonColor,
			buttonTextColor,
			primaryTextColor,
			cssClasses,
		},
		previewField: "data",
	},
};
</script>
<script setup lang="ts">
import { FieldCategory, FieldType } from "@/writerTypes";
import injectionKeys from "@/injectionKeys";
import { computed, inject, readonly, ref, watch } from "vue";
import chroma, { Color } from "chroma-js";
import { useFieldValueAsYesNo } from "@/composables/useFieldValue";
import BaseEmptiness from "../base/BaseEmptiness.vue";
import BaseMarkdownRaw from "../base/BaseMarkdownRaw.vue";

type AnnotatedTextElementArray = [content: string, tag: string, color?: string];
type AnnotatedTextElement = string | AnnotatedTextElementArray;

const fields = inject(injectionKeys.evaluatedFields);

const text = computed<AnnotatedTextElement[]>(() => fields.text.value ?? []);

const isBeingEdited = inject(injectionKeys.isBeingEdited);
const componentId = inject(injectionKeys.componentId);
const isEmpty = computed(() => !fields.text.value);
const shouldDisplay = computed(() => !isEmpty.value || isBeingEdited.value);
const useMarkdown = useFieldValueAsYesNo(fields, "useMarkdown");

const COLOR_STEPS = [
	{ h: -78, s: -34, l: 16 },
	{ h: -61, s: -37, l: 16 },
	{ h: -2, s: 0, l: 24 },
	{ h: -12, s: 0, l: 29 },
	{ h: 28, s: -20, l: 24 },
	{ h: -61, s: -95, l: 25 },
	{ h: -173, s: 0, l: 16 },
	{ h: -228, s: 0, l: 22 },
	{ h: 69, s: 0, l: 25 },
	{ h: 70, s: 0, l: 29 },
];

let currentSteps = [...COLOR_STEPS];
let subjectColorCache = {};
let lastSeed = fields.seed.value;

function useMarkdownRenderer() {
	const markdown = ref("");

	watch(
		[useMarkdown, text],
		async () => {
			if (!useMarkdown.value) return (markdown.value = "");
			markdown.value = await parseMarkdown();
		},
		{ immediate: true },
	);

	/* Translate annotation as `<span>` to use it inside the markdown */
	function RawAnnotation(element: AnnotatedTextElementArray) {
		const [content, subject] = element;
		const subjectEl = subject
			? `<span class="CoreAnnotatedText__annotation__subject">${subject}</span>`
			: "";

		return `<span class="CoreAnnotatedText__annotation" style="background: ${getAnnotationBgColor(element)}">${content}${subjectEl}</span>`;
	}

	async function parseMarkdown() {
		const rawMarkdown = text.value.reduce<string>((acc, part) => {
			if (typeof part === "string") {
				return `${acc} ${part}`;
			} else if (Array.isArray(part)) {
				return `${acc} ${RawAnnotation(part)}`;
			} else {
				return acc;
			}
		}, "");

		const marked = await import("marked");
		return await marked.parse(rawMarkdown);
	}

	return readonly(markdown);
}

const markdown = useMarkdownRenderer();

function generateColorCss(
	baseColor: Color,
	colorData: { h: number; s: number; l: number },
) {
	let genColor = baseColor
		.set(
			"hsl.h",
			`${Math.sign(colorData.h) == -1 ? "-" : "+"}${Math.abs(colorData.h)}`,
		)
		.set(
			"hsl.s",
			`${Math.sign(colorData.s) == -1 ? "-" : "+"}${Math.abs(colorData.s / 100.0)}`,
		)
		.set(
			"hsl.l",
			`${Math.sign(colorData.l) == -1 ? "-" : "+"}${Math.abs(colorData.l / 100.0)}`,
		);

	return genColor.css();
}

function calculateColorStep(s: string, stepsLength = COLOR_STEPS.length) {
	let hash = (fields.seed.value * 52673) & 0xffffffff;
	for (let i = 0; i < s.length; i++) {
		hash = ((i + 1) * s.charCodeAt(i)) ^ (hash & 0xffffffff);
	}
	const step = Math.abs(hash) % stepsLength;

	return step;
}

function getAnnotationBgColor(content: AnnotatedTextElementArray) {
	return content[2] || generateColor(content[1]);
}

function generateColor(s: string) {
	if (fields.rotateHue.value == "no") {
		return fields.referenceColor.value;
	}

	const baseColor = chroma(fields.referenceColor.value);

	if (lastSeed !== fields.seed.value) {
		currentSteps = [...COLOR_STEPS];
		subjectColorCache = {};
		lastSeed = fields.seed.value;
	}

	if (subjectColorCache[s]) {
		return generateColorCss(baseColor, subjectColorCache[s]);
	}

	// If we run out of colors, reset the list
	if (currentSteps.length === 0) {
		currentSteps = [...COLOR_STEPS];
	}

	const colorStep = calculateColorStep(s, currentSteps.length);
	const colorData = currentSteps[colorStep];

	subjectColorCache[s] = colorData;
	currentSteps.splice(colorStep, 1);

	return generateColorCss(baseColor, colorData);
}

const copyRawContent = computed(() => {
	return text.value.reduce<string>((acc, val) => {
		if (typeof val === "string") {
			return acc + val;
		}

		return acc + val[0];
	}, "");
});

const copyStructuredContent = computed(() => {
	try {
		return JSON.stringify(text.value);
	} catch (e) {
		return text.value.join("");
	}
});
</script>

<style scoped>
.CoreAnnotatedText {
	color: var(--primaryTextColor);
	line-height: 1.8;
}

.CoreAnnotatedText__markdown {
	display: flex;
	flex-direction: column;
	gap: 12px;
}
.CoreAnnotatedText__annotation,
.CoreAnnotatedText__markdown :deep(.CoreAnnotatedText__annotation) {
	flex-direction: row;
	align-items: center;
	background: rgba(255, 164, 33, 0.4);
	border-radius: 8px;
	padding: 2px 8px;
	overflow: hidden;
	line-height: 1;
	vertical-align: middle;
}

.CoreAnnotatedText__annotation__subject,
.CoreAnnotatedText__markdown :deep(.CoreAnnotatedText__annotation__subject) {
	display: inline-flex;
	font-size: 10px;
	margin-left: 14px;
	opacity: 0.5;
	position: relative;
	vertical-align: middle;

	&::after {
		content: "";
		border-left: 1px solid;
		opacity: 0.1;
		position: absolute;
		top: 0px;
		left: -9px;
		height: 10px;
	}
}
</style>
