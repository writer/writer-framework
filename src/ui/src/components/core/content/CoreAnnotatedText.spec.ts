import { describe, expect, it } from "vitest";
import BaseMarkdownRaw from "../base/BaseMarkdownRaw.vue";
import CoreAnnotatedText from "./CoreAnnotatedText.vue";
import VueDOMPurifyHTML from "vue-dompurify-html";
import injectionKeys from "@/injectionKeys";
import { buildMockCore, mockProvides } from "@/tests/mocks";
import { flushPromises, shallowMount } from "@vue/test-utils";
import { ref } from "vue";
import { WdsColor } from "@/wds/tokens";

describe("CoreAnnotatedText", async () => {
	const text = [
		"# This\n\n",
		["**is**", "Verb", "red"],
		" some ",
		["_annotated_", "Adjective"],
		["text", "Noun"],
		". ",
		"## title 2\n\n",
		"And [here](https://google.com)'s paragraph 2",
	];

	it("should render in non-markdown mode", async () => {
		const { core } = buildMockCore();

		const wrapper = shallowMount(CoreAnnotatedText, {
			global: {
				plugins: [VueDOMPurifyHTML],
				provide: {
					...mockProvides,
					[injectionKeys.core as symbol]: core,
					[injectionKeys.isBeingEdited as symbol]: ref(false),
					[injectionKeys.evaluatedFields as symbol]: {
						text: ref(text),
						seed: ref(1),
						useMarkdown: ref("no"),
						rotateHue: ref("yes"),
						referenceColor: ref(WdsColor.Blue5),
						copyButtons: ref("yes"),
					},
				},
			},
		});

		await flushPromises();

		const annotations = wrapper.findAll(".CoreAnnotatedText__annotation");
		expect(annotations).toHaveLength(text.filter(Array.isArray).length);

		// should use the value provided
		expect(annotations.at(0).attributes().style).toBe(
			"background-color: red;",
		);

		// should generate the color
		expect(annotations.at(1).attributes().style).toMatchInlineSnapshot(
			`"background-color: rgb(180, 237, 238);"`,
		);

		expect(wrapper.element).toMatchSnapshot();
	});

	it("should render in markdown mode", async () => {
		const { core } = buildMockCore();

		const wrapper = shallowMount(CoreAnnotatedText, {
			global: {
				plugins: [VueDOMPurifyHTML],
				provide: {
					...mockProvides,
					[injectionKeys.core as symbol]: core,
					[injectionKeys.isBeingEdited as symbol]: ref(false),
					[injectionKeys.evaluatedFields as symbol]: {
						text: ref(text),
						seed: ref(1),
						useMarkdown: ref("yes"),
						rotateHue: ref("yes"),
						referenceColor: ref(WdsColor.Blue5),
						copyButtons: ref("yes"),
					},
				},
			},
		});

		await flushPromises();

		expect(
			wrapper.getComponent(BaseMarkdownRaw).props().rawMarkdown,
		).toMatchSnapshot();
	});
});
