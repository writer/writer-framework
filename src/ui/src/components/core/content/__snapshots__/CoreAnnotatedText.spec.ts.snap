// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CoreAnnotatedText > should render in markdown mode 1`] = `""`;

exports[`CoreAnnotatedText > should render in non-markdown mode 2`] = `
<div
  class="CoreAnnotatedText"
  data-v-13ebc398=""
>
  
  <!--v-if-->
  
  <span
    data-v-13ebc398=""
  >
    
    # This


    
    <!--v-if-->
  </span>
  <span
    data-v-13ebc398=""
  >
    <!--v-if-->
    <span
      class="CoreAnnotatedText__annotation"
      data-v-13ebc398=""
      style="background-color: red;"
    >
      **is** 
      <span
        class="CoreAnnotatedText__annotation__subject"
        data-v-13ebc398=""
      >
        Verb
      </span>
    </span>
  </span>
  <span
    data-v-13ebc398=""
  >
    
     some 
    
    <!--v-if-->
  </span>
  <span
    data-v-13ebc398=""
  >
    <!--v-if-->
    <span
      class="CoreAnnotatedText__annotation"
      data-v-13ebc398=""
      style="background-color: rgb(180, 237, 238);"
    >
      _annotated_ 
      <span
        class="CoreAnnotatedText__annotation__subject"
        data-v-13ebc398=""
      >
        Adjective
      </span>
    </span>
  </span>
  <span
    data-v-13ebc398=""
  >
    <!--v-if-->
    <span
      class="CoreAnnotatedText__annotation"
      data-v-13ebc398=""
      style="background-color: rgb(229, 208, 250);"
    >
      text 
      <span
        class="CoreAnnotatedText__annotation__subject"
        data-v-13ebc398=""
      >
        Noun
      </span>
    </span>
  </span>
  <span
    data-v-13ebc398=""
  >
    
    . 
    
    <!--v-if-->
  </span>
  <span
    data-v-13ebc398=""
  >
    
    ## title 2


    
    <!--v-if-->
  </span>
  <span
    data-v-13ebc398=""
  >
    
    And [here](https://google.com)'s paragraph 2
    
    <!--v-if-->
  </span>
  
  
  <shared-control-bar-stub
    copyrawcontent="# This

**is** some _annotated_text. ## title 2

And [here](https://google.com)'s paragraph 2"
    copystructuredcontent="["# This\\n\\n",["**is**","Verb","red"]," some ",["_annotated_","Adjective"],["text","Noun"],". ","## title 2\\n\\n","And [here](https://google.com)'s paragraph 2"]"
    data-v-13ebc398=""
  />
</div>
`;
